### PR TITLE
Extension IDs listed for 'unwantedRecommendations' should be ignored.

### LINKED ISSUE #16: Honor `unwantedRecommendations` setting in .vscode/extensions.json.
Currently, looks like the only way to avoid **Language Support for Java(TM) by Red Hat** recommending **Dependency Analytics** is by disabling all recommendations (setting `extensions.ignoreRecommendations` to `true`).

I think that, if this extension is using VSCode "standard" settings to enable/disable recommendations (once the `java.recommendations.dependency.analytics.show` setting has been removed), it should also respect the `unwantedRecommendations` setting in _.vscode/extensions.json_.

### PR DESCRIPTION
**Pull request by rgrunber**
- Extension IDs listed for 'unwantedRecommendations' should be ignored.
- Fixes #16

### CODE REVIEW

File: src/vscode/impl/recommendationServiceImpl.ts [hunk 1]
```diff
@@ -40,6 +42,18 @@ export class RecommendationServiceImpl implements IRecommendationService {
         return path.resolve(context.globalStorageUri.fsPath, '..', 'vscode-extension-recommender');
     }
 
+    protected getExtensionConfiguration(): string | undefined {
+        if (workspace.workspaceFolders !== undefined) {
+            for (const wf of workspace.workspaceFolders) {
```
Comment (rgrunber) [inline src/vscode/impl/recommendationServiceImpl.ts]: According to https://code.visualstudio.com/docs/editor/extension-marketplace#_workspace-recommended-extensions, in a multi-root workspace, we should check `.code-workspace`.

**Comment by rgrunber** [comment]
@robstryker I think it should be good, but let me know if you see anything odd.

I'll probably have this merged, and take on the long journey of updating all consumers (but probably vscode-java first as the request originates from there).

**Comment by robstryker** [comment]
https://github.com/redhat-developer/vscode-extension-proposals/pull/17/files#diff-edb612d4f1a6f7395a50700ab8b272110e81b8a1d92285aca5b23e54ed62c136R160

This looks like it will read the workspace file `n` times. Surely this can be optimized?

**Comment by rgrunber** [comment]
Added an optional argument for the json data that's read from the configuration file so it can be re-used.

File: src/vscode/impl/recommendationServiceImpl.ts [hunk 2]
```diff
@@ -151,10 +151,19 @@ export class RecommendationServiceImpl implements IRecommendationService {
...
         const model: RecommendationModel|undefined = await this.storageService.readRecommendationModel();        
         if( model ) {
-            const recommendedExtension: string[] = model.recommendations
+            const recommendedExtension: string[] = [];
+            const filteredExtension: string[] = model.recommendations
                 .map((x) => x.extensionId)
                 .filter(filterUnique)
                 .filter((x) => !isExtensionInstalled(x));
+
+            const jsonData = await readExtensionConfigurationFile();
```
Comment (robstryker) [inline src/vscode/impl/recommendationServiceImpl.ts:160]: Much Improved.

**Review by robstryker — APPROVED** [review]

**Pull request merged**